cabby/rvs/templates.py

```diff
@@ -0,0 +1,198 @@
+    Production(Nonterminal('M_G'), (Nonterminal('MAIN'),
+                    Nonterminal('E'))),
+    # End part - (1) near pivot assuming goal already mentioned; and (2) avoid
+    # sentence.
+    Production(Nonterminal('E'), (Nonterminal('NEAR_GOAL_START'),
+                    Nonterminal('AVOID'))),
+    # End part - (1) near pivot assuming goal not mentioned yet; and (2) avoid
+    # sentence.
+    Production(Nonterminal('E'), (Nonterminal('NEAR_GOAL_END'),
+                    Nonterminal('AVOID'))),
+    # Main part of the instruction without verb in begining and resuming
+    # sentence assuming no goal mentioned before.
+    Production(Nonterminal('NO_GOAL'), (Nonterminal('MAIN_NO_V'),
+                      Nonterminal('G'))),
+    # Add Goal to main part and then resume instruction by adding an
+    # ending(near+avoid).
+    Production(Nonterminal('G'), (Nonterminal('GOAL_END'),
+                    Nonterminal('E'))),
+    # Add Goal with near and then add an avoid sentenece.
+    Production(Nonterminal('G'), (Nonterminal('NEAR_GOAL_END'),
+                    Nonterminal('AVOID'))),
+    # Termial for IN+DT after verb.
+    Production(Nonterminal('V1_CON'), ('to the',)),
+
+  ]
+
+  prods += add_rules('V2', V2)
+  prods += add_rules('AVOID', AVOID)
+  prods += add_rules('NEAR_GOAL_START', NEAR_GOAL_START)
+  prods += add_rules('NEAR_GOAL_END', NEAR_GOAL_END)
+  prods += add_rules('GOAL', GOAL)
+  prods += add_rules('GOAL_END', GOAL_END)
+  prods += add_rules('MAIN_NO_V', MAIN_NO_V)
+  prods += add_rules('MAIN', MAIN)
+  prods += add_rules('V1', V1)
+
+  grammar = CFG(Nonterminal('S'), prods)
+
+  # Generate templates.
+  templates = []
+  for sentence in generate(grammar):
+
+    sentence = ' '.join(sentence)
+
+    if sentence[-1] != '.':
+      sentence += '.'
+    sentence = sentence.replace(" .", ".")
+    sentence = sentence.replace(" ,", ",")
+    sentence = sentence.replace("..", ".")
+    sentence = sentence.replace("   ", " ")
+    sentence = sentence.replace("  ", " ")
+    templates.append(sentence)
+
+  templates_df = pd.DataFrame(
+    templates, columns=['sentence']).drop_duplicates()
+  # Save templates
+  templates_df.to_csv('templates.csv', index=False)
+
+  # Flag features.
+  templates_df['cardinal_direction'] = templates_df['sentence'].apply(
+    lambda x: 'CARDINAL_DIRECTION' in x)
+  templates_df['blocks'] = templates_df['sentence'].apply(
+    lambda x: 'NUMBER_BLOCKS' in x)
+  templates_df['intersections'] = templates_df['sentence'].apply(
+    lambda x: 'NUMBER_INTERSECTIONS' in x)
+  templates_df['beyond_pivot'] = templates_df['sentence'].apply(
+    lambda x: 'BEYOND_PIVOT' in x)
+  templates_df['next_block'] = templates_df['sentence'].apply(
+    lambda x: 'next block' in x)
+  templates_df['next_intersection'] = templates_df['sentence'].apply(
+    lambda x: 'next intersection' in x)
+
+  return templates_df
+
+
+def add_features_to_template(template: Text, entity: item.RVSPath) -> Text:
```

Comment:
My previous request to add a test for this seems to have disappeared! (I can see it in the "conversations" but I can't do anything except resolve it... still some weird things about the Github review process...) Anyway, can you please add templates_test.py and test this function? Or at least add an issue and a TODO?